File: src/csharp/Intel/Iced/Intel/BlockEncoderInternal/SimpleInstr.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER
namespace Iced.Intel.BlockEncoderInternal {
	/// <summary>
	/// Simple instruction that doesn't need fixing, i.e., it's not IP relative (no branch instruction, no IP relative memory operand)
	/// </summary>
	sealed class SimpleInstr : Instr {
		Instruction instruction;

		public SimpleInstr(BlockEncoder blockEncoder, Block block, in Instruction instruction)
			: base(block, instruction.IP) {
			Done = true;
			this.instruction = instruction;
			Size = blockEncoder.GetInstructionSize(instruction, instruction.IP);
		}

		public override void Initialize(BlockEncoder blockEncoder) {}
		public override bool Optimize(ulong gained) => false;

		public override string? TryEncode(Encoder encoder, out ConstantOffsets constantOffsets, out bool isOriginalInstruction) {
			isOriginalInstruction = true;
			if (!encoder.TryEncode(instruction, IP, out _, out var errorMessage)) {
				constantOffsets = default;
				return CreateErrorMessage(errorMessage, instruction);
			}
			constantOffsets = encoder.GetConstantOffsets();
			return null;
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/enc/SimpleInstr.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.enc;

import com.github.icedland.iced.x86.Instruction;

final class SimpleInstr extends Instr {
	private Instruction instruction;

	SimpleInstr(BlockEncoder blockEncoder, Block block, Instruction instruction) {
		super(block, instruction.getIP());
		done = true;
		this.instruction = instruction;
		size = blockEncoder.getInstructionSize(instruction, instruction.getIP());
	}

	@Override
	void initialize(BlockEncoder blockEncoder) {
	}

	@Override
	boolean optimize(long gained) {
		return false;
	}

	@Override
	String tryEncode(Encoder encoder, TryEncodeResult result) {
		result.isOriginalInstruction = true;
		Object encResult = encoder.tryEncode(instruction, ip);
		if (encResult instanceof Integer) {
			result.constantOffsets = encoder.getConstantOffsets();
			return null;
		}
		return createErrorMessage((String)encResult, instruction);
	}
}
